selfie.c

```diff
@@ -1150,6 +1150,144 @@ uint64_t debug_switch = 0;
+
+void set_valid_flag(uint64_t* cache_block, uint64_t valid)     { *cache_block       = valid; }
+void set_tag(uint64_t* cache_block, uint64_t tag)              { *(cache_block + 1) = tag; }
+void set_data(uint64_t* cache_block, uint64_t* data)           { *(cache_block + 2) = (uint64_t) data; }
+void set_timestamp(uint64_t* cache_block, uint64_t timestamp)  { *(cache_block + 3) = timestamp; }
+
+void reset_cache_counters(uint64_t* cache);
+void reset_all_cache_counters();
+
+void init_cache_memory(uint64_t* cache);
+void init_cache(uint64_t* cache, uint64_t cache_size, uint64_t associativity, uint64_t cache_block_size);
+void init_all_caches();
+
+void flush_cache(uint64_t* cache);
+void flush_all_caches();
+
+uint64_t  calculate_tag(uint64_t* cache, uint64_t* paddr);
+uint64_t* calculate_set(uint64_t* cache, uint64_t vaddr);
+uint64_t  get_new_timestamp(uint64_t* cache);
+uint64_t* cache_lookup(uint64_t* cache, uint64_t* paddr, uint64_t vaddr, uint64_t is_access);
+void      fill_cache_block(uint64_t* cache_block, uint64_t cache_block_size, uint64_t* start_paddr);
+uint64_t* handle_cache_miss(uint64_t* cache, uint64_t* cache_block, uint64_t* paddr, uint64_t is_access);
+uint64_t* retrieve_cache_block(uint64_t* cache, uint64_t* paddr, uint64_t vaddr, uint64_t is_access);
+
+uint64_t calculate_word_offset(uint64_t cache_block_size, uint64_t vaddr);
+void     flush_cache_block(uint64_t* cache_block, uint64_t cache_block_size, uint64_t* start_paddr);
+void     store_in_cache(uint64_t* cache, uint64_t* paddr, uint64_t vaddr, uint64_t data);
+uint64_t load_from_cache(uint64_t* cache, uint64_t* paddr, uint64_t vaddr);
+
+uint64_t load_instruction_from_cache(uint64_t* paddr, uint64_t vaddr);
+uint64_t load_data_from_cache(uint64_t* paddr, uint64_t vaddr);
+void     store_data_in_cache(uint64_t* paddr, uint64_t vaddr, uint64_t data);
+
+void print_cache_statistic(uint64_t hits, uint64_t misses, char* cache_name);
+
+// ------------------------ GLOBAL CONSTANTS -----------------------
+
+// indicates whether the machine has a cache or not
+uint64_t L1_CACHE_ENABLED = 0;
+
+// machine-enforced coherency for self-modifying code (selfie also
+// works if this is turned off since there is no code modification
+// during runtime and stores in the code segment are illegal)
+uint64_t L1_CACHE_COHERENCY = 0;
+
+// example configurations:
+// +-------------------+---------------+-----------------------------+------------+
+// |              name |    cache size |               associativity | block size |
+// +===================+===============+=============================+============+
+// |       CORE-V CVA6 | dcache: 32 KB |                   dcache: 8 |       16 B |
+// |                   | icache: 16 KB |                   icache: 4 |            |
+// +-------------------+---------------+-----------------------------+------------+
+// |     direct-mapped |         2^x B |                           1 |      2^y B |
```

Comment:
What is `x` and `y` actually?

Reply:
Those are just some arbitrary exponents (i.e. `^` does not represent the bitwise XOR operator here) since we wanted to express that those values are always powers of 2.